hail/src/main/scala/is/hail/io/fs/GoogleStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+package is.hail.io.fs
+
+import java.io.{ByteArrayInputStream, FileNotFoundException, InputStream, OutputStream}
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file.FileSystems
+
+import is.hail.utils._
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.cloud.{ReadChannel, WriteChannel}
+import com.google.cloud.storage.Storage.BlobListOption
+import com.google.cloud.storage.{Blob, BlobId, BlobInfo, Storage, StorageOptions}
+import org.apache.commons.io.FilenameUtils
+import org.apache.hadoop
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object GoogleStorageFS {
+  def containsWildcard(path: String): Boolean = {
+    var i = 0
+    while (i < path.length) {
+      var c = path(i)
+      if (c == '\\') {
+        i += 1
+        if (i < path.length)
+          i += 1
+        else
+          return false
+      } else if (c == '*' || c == '{' || c == '?')
```

Comment:
Looking at: https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames, I don't see "{" in the list, but there may be a more complete description. I think we may also want (\[\w+\]) and (\[a-zA-Z\-a-zA-Z\]\).

This could be made more terse and precise:
```scala
val d = raw"(?<=\/)(\[\w\-\w\]|\[\w+\]|[\*]+|[\?]+)".r
print(d.findFirstIn("/??"))
```

or to allow -, . and _ in character ranges (haven't double checked that this allowed in wildcard, but those are allowable characters): `val d = raw"(?<=\/)(\[[\w\_\-\.]{1}\-[\w\_\-\.]{1}\]|\[[\w\_\.]+\]|[\*]+|[\?]+)".r`

Reply:
Sorry, I meant to write a note about wildcards.

We already support and document the Hadoop wildcard syntax: https://hail.is/docs/0.2/hadoop_glob_patterns.html?highlight=hadoop via the Hadoop library.  I looked at the Hadoop code, and it wasn't possible to use externally.  The closest thing I could find is the Java NIO glob syntax: https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String).  I think the main difference is that the Java syntax uses ^ instead of ! for character negation, and doesn't support {} which Hadoop does, but it was easy to use.

So propose we take the Java version for now, but before it becomes user facing, either implement our own version of the Hadoop rules (or a superset) and use that.

Another option is to have the glob rules be backend-specific.  I think that is a mistake.

Another option is to change the glob syntax (e.g. everywhere Java).  That will require finishing ripping out the other uses of Hadoop and writing our own HadoopFS.glob.  That isn't so bad, but I think we should avoid breaking the existing interface.

Finally, I am relaxing the definition of FileStatus (and the return type of hadoop_stat) because e.g.Google doesn't have a natural owner, so I just return None in that case.  For this, I think we should try to document the schema per-backend, and choose common names when possible.

Reply:
> Another option is to have the glob rules be backend-specific. I think that is a mistake.

Agreed, too much cognitive overhead to users, no gain.

> For this, I think we should try to document the schema per-backend, and choose common names when possible.

Sure, as we saw with gcsfs, perfect compatibility isn't possible, google buckets are not file systems. Stuff like last modification date can be emulated, permissions 


Thanks for the explanation. My original point was that you rolled a 20 line function that doesn't enforce these rules precisely. This can be done in 1 line of regex "(\* | \{ | \? )". However, after reading your explanation I think this function should be dropped altogether. `val m = javaFS.getPathMatcher(s"glob:$c")` and check if `m.matches(path)`. This will be less fragile.

Reply:
Ah I see, you need to worry about trailing slashes, the matcher syntax is more specific. In that case, I'd prefer the regex solution, but if you feel strongly about your version, drop the "\\", which is unnecessary.

Reply:
> Is the matcher syntax too specific to match across trailing slashes?

I'm not sure I understand the question.

I need to handle the backslahses because "foo\*bar" needs to match the file "foo*bar" and not, say, "fooextrabar".

The optimization here is that I don't need to do a directory listing (saving a round trip to gs://) if the file does not contain a wildcard.

Reply:
Yes, need to add [, good catch.